snapcraft/internal/elf.py

```diff
@@ -301,14 +306,17 @@ def __init__(self, *, dynamic_linker: str, root_path: str) -> None:
         # in by packaging dependencies.
         # The docker conditional will work if the docker image has the
         # snaps unpacked in the corresponding locations.
-        if common.is_snap():
+        which_patchelf = shell_utils.which('patchelf')
+        if which_patchelf.endswith('stage/patchelf'):
+            self._patchelf_cmd = which_patchelf
+        elif common.is_snap():
             snap_dir = os.getenv('SNAP')
             self._patchelf_cmd = os.path.join(snap_dir, 'bin', 'patchelf')
         elif (common.is_docker_instance() and
               os.path.exists('/snap/snapcraft/current/bin/patchelf')):
             self._patchelf_cmd = '/snap/snapcraft/current/bin/patchelf'
         else:
-            self._patchelf_cmd = 'patchelf'
+            self._patchelf_cmd = which_patchelf
```

Comment:
This is the same check as the first if, so you can rewrite this statement to something like.
```
self._patchelf_cmd = shell_utils.which('patchelf')
if not patchelf in stage:
    if is_snap():
        [...]
    elif docker:
        [...]
```
Looks more readable to me.